### PR TITLE
adds translator for dreier informationsdienst

### PR DESCRIPTION
**Pull request by denismaier**
This adds a translator for the book supplier *Dreier Informationsdienst*.

For some reasons the cases don't work currently. Maybe because the page needs a login?

Edit: It works now after removing the tests. These were:

```javascript
/** BEGIN TEST CASES **/
var testCases = [
	{
		"type": "web",
		"url": "https://www.dietmardreier.de/de/detail/ISBN-9781792457418/Suehr-Christopher-J./A-Firm-Foundation",
		"items": [
			{
				"itemType": "book",
				"title": "A Firm Foundation: A Simple Introduction to the Christian Bible",
				"creators": [
					{
						"firstName": "Christopher J.",
						"lastName": "Suehr",
						"creatorType": "author"
					}
				],
				"date": "2020",
				"ISBN": "9781792457418",
				"abstractNote": "The Bible is big, complicated, and one of the most important books in human history.A Firm Foundation introduces the book that many call sacred, but few have studied.This versatile resource presents the Bible in plain language, with straightforward summaries, questions to facilitate discussion and reflection, and clearly defined terminology. It can stand alone, but works best as a companion to the bible.The material within has been used to teach a variety of students, including nonreligious beginners, clergy-in-training, English-language learners, and followers of other major world religions.Build a stronger understanding with A Firm Foundation.",
				"libraryCatalog": "Dreier Neuerscheinungsdienst",
				"numPages": "277 Seiten",
				"place": "Iowa",
				"publisher": "Kendall/Hunt Publishing Co ,U.S.Kendall",
				"shortTitle": "A Firm Foundation",
				"attachments": [],
				"tags": [],
				"notes": [],
				"seeAlso": []
			}
		]
	},
	{
		"type": "web",
		"url": "https://www.dietmardreier.de/de/detail/ISBN-9781626168084/Mosher-Lucinda/A-World-of-Inequalities",
		"items": [
			{
				"itemType": "book",
				"title": "A World of Inequalities: Christian and Muslim Perspectives",
				"creators": [
					{
						"firstName": "Lucinda",
						"lastName": "Mosher",
						"creatorType": "editor"
					}
				],
				"date": "2021",
				"ISBN": "9781626168084",
				"abstractNote": "In this volume, leading Christian and Muslim scholars respond to the global crisis of inequality by demanding and modeling interreligious dialogue. Essays explore the roots of these realities, how they are treated in Christian and Muslim traditions and texts, and how the two faiths can work together to address inequality.",
				"libraryCatalog": "Dreier Neuerscheinungsdienst",
				"numPages": "256 Seiten",
				"place": "Washington, DC",
				"publisher": "Georgetown University Press",
				"shortTitle": "A World of Inequalities",
				"attachments": [],
				"tags": [],
				"notes": [],
				"seeAlso": []
			}
		]
	},
	{
		"type": "web",
		"url": "https://www.dietmardreier.de/de/detail/ISBN-9783806240580/Nussbaum-Martha/Kosmopolitismus",
		"items": [
			{
				"itemType": "book",
				"title": "Kosmopolitismus: Revision eines Ideals",
				"creators": [
					{
						"firstName": "Martha",
						"lastName": "Nussbaum",
						"creatorType": "author"
					},
					{
						"firstName": "Manfred",
						"lastName": "Weltecke",
						"creatorType": "translator"
					}
				],
				"date": "2020",
				"ISBN": "9783806240580",
				"abstractNote": "Was bedeutet die Würde des Menschen? Ein philosophischer Essay Der kynische Philosoph Diogenes bezeichnete sich selbst einen kosmopolitês, einen \"Bürger der Welt\". Damit war der Kosmopolitismus geboren. Ein Kosmopolit definiert sich als Mensch - unabhängig von seiner Abstammung, seinem Geschlecht oder seiner sozialen Herkunft. Diese Haltung ist die Geburtsstunde der Menschenwürde in der abendländischen Kultur. Die amerikanische Philosophin Martha Nussbaum erörtert in ihrem Essay die kulturelle Tradition des Kosmopolitismus in Europa seit der Antike. Sie erläutert die philosophische Idee grundlegend und zeigt Querverbindungen zu aktuellen gesellschaftlichen Debatten auf. Entwicklung der Idee einer universellen Menschenwürde von der Antike bis in die Gegenwart Ethische Dilemmas aus Pluralismus & Globalisierung: Kosmopolitismus als Lösung? Weiterentwicklung der philosophischen Idee: der Fähigkeitenansatz von Martha Nussbaum Tiefgründig und inspirierend: Die Autorin gilt als einflussreichste Philosophin der Gegenwart Die Grundlagen herausfordern: Viele Wesen, viele Arten von Würde Kann in Zeiten des Globalismus und Pluralismus der Kosmopolitismus eine Lösung für die aktuellen politisch-gesellschaftlichen Herausforderungen sein? Auf der Suche nach der Antwort untersucht Martha Nussbaum in ihrem Sachbuch die philosophische Strömung von den griechischen Stoikern und Cicero über Hugo Grotius, Immanuel Kant bis zu Adam Smith. Sie analysiert so den Einfluss des Kosmopolitismus in der politischen Theorie und benennt konsequent die Stärken und Defizite. Mit dem von ihr mitentwickelten Fähigkeitenansatz erweitert sie zudem den modernen Kosmopolitismus radikal über die menschliche Spezies hinaus. Eine tiefgründige Analyse sowohl für Philosophen als auch für politisch Interessierte: Dieses Buch regt zum Nachdenken an!",
				"libraryCatalog": "Dreier Neuerscheinungsdienst",
				"numPages": "352 Seiten",
				"place": "Darmstadt",
				"publisher": "wbg Theiss in Wissenschaftliche Buchgesellschaft (WBG)",
				"shortTitle": "Kosmopolitismus",
				"attachments": [],
				"tags": [],
				"notes": [],
				"seeAlso": []
			}
		]
	}
]
/** END TEST CASES **/
```

Can anyone tell what's wrong with these?

### CODE REVIEW

**Comment by Mailaender** [comment]
> For some reasons the cases don't work currently. Maybe because the page needs a login?

Yes.

**Comment by denismaier** [comment]
@Mailaender Ok, thanks for the clarification. So, no way to add tests in this case?

**Comment by Mailaender** [comment]
You will need to find a page that is not password protected.

File: Dreier Neuerscheinungsdienst.js [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+	else if (getSearchResults(doc, true)) {
+		return "multiple";
+	}
+	return false;
+}
+
+function getSearchResults(doc, checkOnly) {
+	var items = {};
+	var found = false;
+	// TODO: adjust the CSS selector
```
Comment (AbeJellinek) [inline Dreier Neuerscheinungsdienst.js:52]: can be removed

File: Dreier Neuerscheinungsdienst.js [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+	var rows = doc.querySelectorAll('h4 a');
```
Comment (AbeJellinek) [inline Dreier Neuerscheinungsdienst.js:53]: ...but maybe make this more specific (is there a class on the `h4` or `a` that we could target?)

File: Dreier Neuerscheinungsdienst.js [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+	}
+}
+
+function scrape(doc) {
+	var item = new Zotero.Item("book");
+	// Set the title
+	item.title = ZU.xpathText(doc, '//h1[@class="biblioTitle"]');
+	if (ZU.xpathText(doc, '//div[@class="titles"]/div[@class="biblioSubTitle"]')) item.title += ": " + ZU.xpathText(doc, '//div[@class="titles"]/div[@class="biblioSubTitle"]');
+	// ISBN
+	item.ISBN = (doc.querySelector('.biblioId .value').textContent) ? doc.querySelector('.biblioId .value').textContent : "";
```
Comment (AbeJellinek) [inline Dreier Neuerscheinungsdienst.js:82]: All of these `textContent` ternaries can be simplified to calls to `text`. For instance, this one could be

```js
item.ISBN = text(doc, '.biblioId .value');
```

File: Dreier Neuerscheinungsdienst.js [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+	item.abstractNote = (doc.querySelector('.description .blurb .value').textContent) ? doc.querySelector('.description .blurb .value').textContent : "";
+	// Get Creators
+	var creators = doc.querySelectorAll('.authorMain .biblioAuthor');
+	for (let creator of creators) {
+		let creatorName = creator.querySelector('.value').textContent;
+		let creatorRole;
+		
+		// currently, we only check for editors; everything else will be treated as authors
+		if (creator.textContent.includes("Hrsg.")) {
+			Z.debug(creatorName + " ist Herausgeber");
```
Comment (AbeJellinek) [inline Dreier Neuerscheinungsdienst.js:101]: I know this translator targets a German publisher and will probably mostly have German users, but in the spirit of the code and comments being in English, I think the debug messages should be, too.

**Comment by AbeJellinek** [comment]
It's alright to have tests that only work locally - IMO better than not having them at all. You'll just need to confirm that they *do* work locally. If you add them back and address my inline comments, I'll be happy to merge this!

**Comment by denismaier** [comment]
> It's alright to have tests that only work locally - IMO better than not having them at all. You'll just need to confirm that they _do_ work locally.

Hmm, I have to admit I'm a bit lost here. I've added the tests back in, but I'm always getting the error message "detection failed".  Strange thing is: if I copy the url to the browser window and click on "run detectWeb" and "run doWeb", everything works as expected. Any hints on that?

File: Dreier Neuerscheinungsdienst.js (hunk 3 above)
Comment (denismaier) [reply Dreier Neuerscheinungsdienst.js]: Thanks for the hint. Done.

File: Dreier Neuerscheinungsdienst.js (hunk 4 above)
Comment (denismaier) [reply Dreier Neuerscheinungsdienst.js]: I've removed the debug messages entirely. I've also made the tests more robust by adding French and English terms. Now creators work now also in the English and French version of the webpage.

**Comment by AbeJellinek** [comment]
Yeah, seems to be something about the login redirect. Oh well. If the tests reflect the output on your end, I'm happy with it. (I'd like to run tests locally here, but I don't have an account...)

**Comment by denismaier** [comment]
Thanks for merging this. What will happen next? When will this be available in Zotero? (Does this require a Zotero update or are translator updates independent of Zotero updates?

**Comment by adam3smith** [comment]
Translators update almost immediately, 24hs at most (you can also trigger update from the prefs). New ones may require a browser restart or updating translators from the connector.

**Pull request merged**